storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -2812,6 +2816,76 @@ fseg_free_extent(
 	return DB_SUCCESS;
 }
 
+/** Free the extent and fragment page associated with
+the segment.
+@param iblock     page where segment header are placed
+@param inode      index node information
+@param space      tablespace where segment resides
+@param mtr        mini-transaction
+@param header     segment header
+@param free_inode free the inode page also
+@param ahi        adaptive hash index
+@return DB_SUCCESS_LOCKED_REC when freeing wasn't completed
+@return DB_SUCCESS or other error code  when freeing was completed */
+static
+dberr_t fseg_free_step_low(buf_block_t *iblock, fseg_inode_t *inode,
+                           fil_space_t *space, mtr_t *mtr,
+                           fseg_header_t *header, bool free_inode
+#ifdef BTR_CUR_HASH_ADAPT
+                           , bool ahi=false
+#endif /* BTR_CUR_HASH_ADAPT */
+                           )
+{
+  dberr_t err= DB_SUCCESS;
+  if (xdes_t *descr= fseg_get_first_extent(inode, space, mtr, &err))
+  {
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), mtr
+#ifdef BTR_CUR_HASH_ADAPT
+                          , ahi
+#endif /* BTR_CUR_HASH_ADAPT */
+                        );
+    if (err == DB_SUCCESS)
+      err= DB_SUCCESS_LOCKED_REC;
+  }
+
+  if (err != DB_SUCCESS)
+    return err;
+
+  /* Free a frag page */
+  ulint n = fseg_find_last_used_frag_page_slot(inode);
+  if (UNIV_UNLIKELY(n == ULINT_UNDEFINED))
+  {
+    if (free_inode)
+      fsp_free_seg_inode(space, inode, iblock, mtr);
+    return DB_SUCCESS;
+  }
+
+  uint32_t page_no = fseg_get_nth_frag_page_no(inode, n);
+  if (header && page_no == page_get_page_no(page_align(header)))
+     return DB_SUCCESS;
```

Comment:
Should there be a `.` after the first line of the comment? There’s also a grammar error "should not freed".

This function is called by 3 functions. I see that there is a third possible caller `fseg_inode_free()`, but I don’t think we need to mention that in the comment. I would place the comment between the `if` and the `return` (which by the way is indented 1 column too deep), like this. Furthermore, I would name the parameter `const byte *header_page_no` so that we will avoid calling `page_align(header)` twice (once here and another time in our caller):

(We can compare the 4 bytes directly.)
```c++
  if (header_page_no && !memcmp_aligned<4>(header_page_no,
                                   inode + FSEG_FRAG_ARR +
                                   n * FSEG_FRAG_SLOT_SIZE, 4))
    /* header_page_no is only passed by fseg_free_step_not_header().
    In that case, the header page must be preserved, to be freed when we're
    finally called by fseg_free_step(). */
    return DB_SUCCESS;
```
Furthermore, the parameter `bool free_inode` is redundant and should be replaced by `!header_page_no`.